tests/test_checkpointing.py

```diff
@@ -0,0 +1,122 @@
+from tests.fixtures.kafka import KAFKA_SERVICE_NAME, KAFKA_PACKAGE_NAME
+from tests.fixtures.kafka import get_kerberized_kafka_spark_conf, upload_jaas
+
+LOGGER = logging.getLogger(__name__)
+HDFS_CHECKPOINT_DIR = "hdfs://hdfs/checkpoints"
+HDFS_TMP_DIR = "/tmp"
+KAFKA_TEST_TOPIC = "streaming_test"
+SPARK_APPLICATION_NAME = "StructuredStreamingWithCheckpointing"
+SPARK_SECURITY_PROTOCOL = "SASL_PLAINTEXT"
+
+def setup_hdfs_paths():
+    sdk_auth.kinit(HDFS_CLIENT_ID, keytab="hdfs.keytab", principal=GENERIC_HDFS_USER_PRINCIPAL)
+    hdfs_cmd("rm -r -skipTrash {}".format(HDFS_CHECKPOINT_DIR))
+    hdfs_cmd("mkdir {}".format(HDFS_CHECKPOINT_DIR))
+    hdfs_cmd("chmod 1777 {}".format(HDFS_CHECKPOINT_DIR))
+
+    hdfs_cmd("rm -r -skipTrash {}".format(HDFS_TMP_DIR))
+    hdfs_cmd("mkdir {}".format(HDFS_TMP_DIR))
+    hdfs_cmd("chmod 1777 {}".format(HDFS_TMP_DIR))
+
+
+def feed_sample_data(jar_uri, kafka_brokers, topic, common_args, messages):
+    producer_args = ["--class", "KerberizedKafkaProducer"] + common_args
+    producer_id = utils.submit_job(app_url=jar_uri,
+                                   app_args="{} {} {} {}".format("kafka", kafka_brokers, topic, ' '.join(messages)),
+                                   service_name=utils.SPARK_SERVICE_NAME,
+                                   args=producer_args)
+
+    # validating producer output
+    utils.check_job_output(producer_id, "{} messages sent to Kafka".format(len(messages)))
+
+
+@pytest.mark.skipif(not utils.hdfs_enabled(), reason='HDFS_ENABLED is false')
+@pytest.mark.sanity
+def test_structured_streaming_recovery(kerberized_spark, kerberized_kafka):
+
+    @retrying.retry(
+        wait_fixed=5000,
+        stop_max_delay=600 * 1000,
+        retry_on_result=lambda res: not res)
+    def wait_for_running_job_output(task_id, expected_line):
+        stdout = sdk_cmd.run_cli("task log --lines=1000 {}".format(task_id))
+        result = expected_line in stdout
+        LOGGER.info('Checking for {} in STDOUT:\n{}\nResult: {}'.format(expected_line, stdout, result))
+        return result
+
+    kafka_brokers = ','.join(sdk_cmd.svc_cli(KAFKA_PACKAGE_NAME, KAFKA_SERVICE_NAME, 'endpoints broker', json=True)['dns'])
+    LOGGER.info("Kafka brokers: {}".format(kafka_brokers))
+
+    _uri = upload_jaas()
+    uris = "spark.mesos.uris={}".format(_uri)
+
+    jar_uri = utils.dcos_test_jar_url()
+
+    kafka_kerberos_args = get_kerberized_kafka_spark_conf(utils.SPARK_SERVICE_NAME)
+    LOGGER.info("Spark Kerberos configuration for Kafka:\n{}".format('\n'.join(kafka_kerberos_args)))
+
+    common_args = [
+        "--conf", "spark.mesos.containerizer=mesos",
+        "--conf", "spark.scheduler.maxRegisteredResourcesWaitingTime=2400s",
+        "--conf", "spark.scheduler.minRegisteredResourcesRatio=1.0",
+        "--conf", uris
+    ] + kafka_kerberos_args
+
+    # configuring streaming job and HDFS folders
+    setup_hdfs_paths()
+
+    # running kafka producer
+    messages = ["test"] * 100
```

Comment:
Would it make sense to make these messages unique for a specific run?:
`["test0", "test1", ... ]`?

 ```suggestion
     messages = ["test%d" % d for d in range(100)]
```

Reply:
Good catch! The new implementation uses two different word sets to verify both the count accuracy and aggregated data checkpointing.